### ISSUE TITLE
Brilliant ! Making things isomorphic would make it shine !

### ISSUE THREAD

**Issue by paulvanbladel** [beeman/loopback-angular-admin#32]
Dear Bram,

Really great approach and I'm still exploring. I've done some research in strongloop/loopback and I'm really convinced they offer a rock solid foundation for all types of business application (or call it line of business apps).

The real beauty could get in when going the full isomorphic path. 
I'm seeing currently a few examples:
1. Validation logic : usually validation logic client side is a subset of the server side validation logic. If would be cool if the logic is organised in such a way it's written only once and the server reuses the code file of the client side  logic.
2. entity construction (instantiation): in business apps an entity constructor makes sense for providing default values (in such a way they don't need to be set in the controller in a screen specific manner. Often the same entities are created also server side. So, a candidate for client/server sharing.
3. permissions: typical _CanExecute behavior.

Just some random thoughts about the usage angular-formly. I'm not convinced it will be a big time saver. In the end, you will want to set specific css on controls. What about auto complete drop down, more complex foreign key and object composition/aggregation.  I know that typing the html can be tedious, but a good editor with zen coding and smart snippets (as in webstorm) can work as well. 
Things would be different if you would create the formly screens based on the info coming from /common/models, but then you would be limited to pure CRUD screens (i.e. no flexibility in the shape of the entities).

Great work !!!
Kind regards
paul.

**Comment by beeman** [comment]
Hello Paul,

Thanks a lot for your kind words regarding this project! I also think that Strongloop offers some great software, Loopback has already saved me a lot of time while playing around with REST api's and related stuff.

I fully agree with the examples that you gave. I think there will be a lot of gain when the Loopback models are getting more intelligent, and this is something that is needed to make this project ready for production.

I have been thinking about a way to send the metadata about the models to the client. In that way it would be more easy to share validation logic, and it could also aid in automagically creating the forms (with or without angular-formly). For the time being the latter will stay included, because it saves some time in creating forms, and there's always an option to ignore formly and code the forms manually.

If you have any code that you want to share, please send in a pull request! If you'd like to discuss the functionality in more details we could have a chat, for example in the [Gitter chat room](https://gitter.im/beeman/loopback-angular-admin).
 
Thanks for your input! 

Regards, Bram

**Comment by paulvanbladel** [comment]
Hi Bram,
Yeah, for pure line of business (close to Crud or with medium complexity) there is definitely a need for:
a) generating the 'skeleton' of screens in such a way the boilerplate can be scaffold-ed  and can be fine-tuned/reorganised. The reason why I'm not that much attracted by formly-alike generators is (apart from the fact you lose 'degrees of freedom' in screen design') because I see thing evolving in the right direction with polymer/material design where you could define kind of 'business' components, which would make the markup really light and expressive (and as such take a way the need for formly and others). But I agree, in your current implementation, as a developer, you can simply ignore formly, if you don't like it :)
I like also the idea of a server model which is propagated towards the client. Preferably the model logic should be client side technology (e.g. angular) agnostic and client-side specific 'adapters' could 'inject' the model logic in a way compatible with the client framework. That would keep your current implementation open for combining strongloop with other client techno.
b) most line of business apps need entities in a two fold way, either 
(1) in the 'pure' form, the way how the database likes them (a customer, an order),  This pure form is often needed for create/edit screens. 
(2) in a more DTO (data transfer object) alike form, most of the times, in list screens (many items) where aggregation of data is shown (simplistic example: a customer list enriched with the total amount of outstanding orders) or 'summary' screens on a single entity. This DTO style is mostly needed for read-only screens.
This type of segregation goes quite well in combination with a **light** variant of the Command Query Responsibility Segregation pattern server side. In Loopback, the 'models' are meant for the above type 1 (so, the 'pure form'). I think it would be nice to have something server side that can accommodate type 2 (the read-only DTO's) as well. If the DTO exist also in the kind of meta data model, they are a direct candidate for screen scaffolding and again a productivity tool.
For the rest loopback is as a serverside framework quite complete. 

We keep in touch.
Thanks a lot
paul.

**Comment by paulvanbladel** [comment]
I'll try to experiment with the above ideas and keep you posted.

**Comment by beeman** [comment]
Thanks for your input Paul, i'm looking forward to see what the design you are talking about looks like, and how it would help mature the project.

Regards, 
Bram
